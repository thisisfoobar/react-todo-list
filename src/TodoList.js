import { useState } from "react";
import { v4 as uuid } from "uuid";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";

const TodoList = () => {
  const INITIAL_STATE = [
    // { id: uuid(), todo: "Walk the cats" },
    // { id: uuid(), todo: "Feed the lawn" },
  ];
  const [todos, setTodos] = useState(INITIAL_STATE);
  const addTodo = (newTodo) => {
    setTodos((todos) => [...todos, { ...newTodo, id: uuid() }]);
  };
  const removeTodo = (id) => {
    setTodos(todos => todos.filter(todo => todo.id !== id))
  }

  return (
    <div>
      <h2>Todo List</h2>
      <NewTodoForm addTodo={addTodo} />
      <div>
        {todos.map(({ id, todo }) => (
          <Todo key={id} id={id} todo={todo} handleRemove={removeTodo} />
        ))}
      </div>
    </div>
  );
};

export default TodoList;
