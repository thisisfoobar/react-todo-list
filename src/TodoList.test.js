import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

it("renders without crashing", function () {
  render(<TodoList />);
});

it("matches the snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

function addTodo(todoList, todo = "Walk the cats") {
  const todoInput = todoList.getByLabelText("Todo")
  fireEvent.change(todoInput, {target: {value: todo}})
  const button = todoList.getByText("Add New Todo")
  fireEvent.click(button)
}

it("can add a new todo", function () {
  const todoList = render(<TodoList />)

  expect(todoList.queryByText("X")).not.toBeInTheDocument();

  addTodo(todoList);

  const removeButton = todoList.getByText("X")
  expect(removeButton).toBeInTheDocument();
  const addedTodo = todoList.getByText("Walk the cats")
  expect(addedTodo).toBeInTheDocument();

  expect(todoList.getAllByDisplayValue("")).toHaveLength(1)
})

it("removes a todo", function() {
  const todoList = render(<TodoList />)
  addTodo(todoList)

  const removeButton = todoList.getByText("X")
  fireEvent.click(removeButton)
  expect(removeButton).not.toBeInTheDocument()
})