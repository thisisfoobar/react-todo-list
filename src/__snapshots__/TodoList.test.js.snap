// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Todo List
    </h2>
    <form>
      <label
        for="todo"
      >
        Todo
      </label>
      <input
        id="todo"
        name="todo"
        placeholder="New Todo"
        type="text"
        value=""
      />
      <button>
        Add New Todo
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
